### PR TITLE
Add ECR repository construct

### PR DESCRIPTION
**Pull request by MatthewCane**
## Describe your changes

Add an ECR repository construct with defaults tailored to the image tagging policy.

## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] I have updated the Readme / docs
- [x] If this is a breaking change, I have discussed the roll-out and roll-back plan with the team(s) and I have provided detailed instructions
- [x] If it is a core feature, I have added thorough tests

### CODE REVIEW

**Comment by MatthewCane** [comment]
I have no idea why the tests are failing in CI, they work perfectly well locally.

**Comment by ruthwells** [comment]
Is it also worth adding the stack output with the command that you need to run for the scan results (see infra-components). a bit like we do for the lwac commands or the helm diffs?

**Comment by MatthewCane** [comment]
Sounds like a reasonable feature!

**Comment by MatthewCane** [comment]
@ruthwells I've added a stack output

**Comment by ruthwells** [comment]
Hi @MatthewCane I was actually thinking about this output [here](https://github.com/citizensadvice/infra-components/blob/develop/resources/templates/cloudformation_ruby_dsl/beanstalk/ecr.rb#L73) which gives you the command to get scan outputs? But maybe that's overkill?

**Comment by MatthewCane** [comment]
Oh I see what you mean @ruthwells! I think that has a place but not here I don't think. As a pre-written Action, absolutely.

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+        additional_accounts_push: list[str] = [],
+        dev_image_max_age: int = 90,
+        max_images: int = 1000,
+        scan_on_push: bool = True,
+        outputs_enabled: bool = True,
+    ):
+        super().__init__(scope, id)
+        lifecycle_rules = []
+
+        lifecycle_rules.append(
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:44]: let's list the rules in priority order (lowest to highest) as this is how they are evaluated

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                rule_priority=19,
+                description="Delete untagged images",
+                tag_status=TagStatus.UNTAGGED,
+                max_image_age=Duration.days(1),
+            )
+        )
+        lifecycle_rules.append(
+            LifecycleRule(
+                rule_priority=18,
+                description="Delete development images after 90 days year",
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:63]: a whole Mercurian year and a bit more!

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+                tag_status=TagStatus.TAGGED,
+                tag_prefix_list=["dev_"],
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:65]: I would the that prefix an instance attribute that defaults to `dev_`

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+        self.repository = Repository(
+            self,
+            "Default",
+            repository_name=name,
+            lifecycle_rules=lifecycle_rules,
+            image_scan_on_push=scan_on_push,
+        )
+
+        for account in additional_accounts_pull:
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:80]: I don't think it's worth having these here as we have direct access to `self.repository` outside the object

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+            self.repository.grant_pull(iam.AccountPrincipal(account))
+
+        for account in additional_accounts_push:
+            self.repository.grant_push(iam.AccountPrincipal(account))
+
+        if outputs_enabled:
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:86]: IMO constructs shouldn't define outputs as the output name will change depending on which stack the construct is used in; also, adding the output is trivial and can be easily done outside of the class

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            LifecycleRule(
+                rule_priority=18,
+                description="Delete development images after 90 days year",
+                tag_status=TagStatus.TAGGED,
+                tag_prefix_list=["dev_"],
+                max_image_age=Duration.days(dev_image_max_age),
+            )
+        )
+
+        lifecycle_rules += additional_lifecycle_rules
```
Comment (zhelyan) [inline ca_cdk_constructs/ecr/ecr_repository.py:70]: no need to do that as we can set them with `EcrRepository(......).repository.add_lifecycle_rule(....)`

**Review by zhelyan — CHANGES_REQUESTED** [review]
see comments

File: ca_cdk_constructs/ecr/ecr_repository.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+    :param scope: The scope of the construct, usually self.
+    :param id: The id of the construct.
+    :param name: The name of the repository. Must be unique within the account.
+    :param additional_lifecycle_rules: Additional lifecycle rules to add in addition to the default rules. Priority must be unique and equal to or less than 10 as not to conflict with existing rules.
+    :param additional_accounts_pull: Additional accounts that are allowed to pull from the repository in the form of their account number
+    :param additional_accounts_push: Additional accounts that are allowed to push to the repository in the form of their account number.
+    :param dev_image_max_age: The maximum age of a development image in days. Defaults to 90.
+    :param max_images: The maximum number of images allowed in the repository. Defaults to 1000.
+    :param scan_on_push: Whether to scan for vulnerabilities on the image on push. Defaults to True.
```
Comment (Shaqil1989) [inline ca_cdk_constructs/ecr/ecr_repository.py:25]: At least the `outputs_enabled` property is missing above, and should the order of all of the `:param` above be the same as the below in the `__init__`

**Comment by MatthewCane** [comment]
@zhelyan & @Shaqil1989 I have addressed all the comments, could I please have a re-review 🙂

**Review by zhelyan — APPROVED** [review]

**Review by zhelyan — CHANGES_REQUESTED** [review]
update the tests. Actually there's no point testing the construct as there's no custom logic at all.

**Review by Shaqil1989 — COMMENTED** [review]
Looks good to me regarding the comments i made before.

**Comment by MatthewCane** [comment]
@zhelyan done!

**Pull request merged**